src/high-level/abstract-tensor.lisp

```diff
@@ -0,0 +1,262 @@
+             (shape source1) (shape source2))
+     nil)
+    (let ((target (or target (copy-tensor source1))))
+      (map-indexes
+       (shape source1)
+       (lambda (&rest dims)
+         (apply #'(setf tref)
+                (cl:+ (apply #'tref source1 dims)
+                      (apply #'tref source2 dims))
+                target dims)))
+      target)))
+
+(defgeneric - (source1 source2 &optional target)
+  (:documentation "Subtract tensors elementwise, optionally storing the result in TARGET.
+If TARGET is not specified then a new tensor is created with the same element type as the first source tensor")
+  (:method ((source1 abstract-tensor) (source2 abstract-tensor) &optional target)
+    (policy-cond:policy-if
+     (< speed safety)
+     (assert (equalp (shape source1) (shape source2))
+             () "Incompatible shapes. Cannot add tensor of shape ~a to tensor of shape ~a."
+             (shape source1) (shape source2))
+     nil)
+    (let ((target (or target (copy-tensor source1))))
+      (map-indexes
+       (shape source1)
+       (lambda (&rest dims)
+         (apply #'(setf tref)
+                (cl:- (apply #'tref source1 dims)
+                      (apply #'tref source2 dims))
+                target dims)))
+      target)))
+
+(defgeneric = (source1 source2 &optional epsilon)
+  (:documentation "Check the equality of tensors with an optional EPSILON")
+  (:method ((source1 abstract-tensor) (source2 abstract-tensor) &optional epsilon)
+    (declare (ignore epsilon))
+    (unless (equal (shape source1) (shape source2))
+      (return-from = nil))
+    (map-indexes
+     (shape source1)
+     (lambda (&rest pos)
+       (unless (equal
+                (apply #'tref source1 pos)
+                (apply #'tref source2 pos))
+         (return-from = nil))))
+    t))
+
+(defgeneric every (predicate tensor)
+  (:documentation "Check that every element in TENSOR meets a given condition")
+  (:method ((predicate function) (tensor abstract-tensor))
+      (foreach (lambda (x)
+                 (unless (funcall predicate x)
+                   (return-from every nil)))
+               tensor)
+    t))
+
+(defgeneric some (predicate tensor)
+  (:documentation "Check that some elements in TENSOR meet a given condition")
+  (:method ((predicate function) (tensor abstract-tensor))
+      (foreach (lambda (x)
+                 (when (funcall predicate x)
+                   (return-from some t)))
+               tensor)
+    nil))
+
+(defgeneric notevery (predicate tensor)
+  (:documentation "Check that not every element in TENSOR meets a given condition")
+  (:method ((predicate function) (tensor abstract-tensor))
+    (not (every predicate tensor))))
+
+(defgeneric notany (predicate tensor)
+  (:documentation "Check that not any element in TENSOR meets a given condition")
+  (:method ((predicate function) (tensor abstract-tensor))
+    (not (some predicate tensor))))
```

Comment:
Seeing these functions now makes me think we might want to play ball with `sb-sequence` in SBCL. This would be work for MAGICL-NG++ 2.01, though.